### PR TITLE
Update Rust Toolchain to Latest Nightly

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR updates our toolchain to `nightly-2021-05-12`.
It depends on https://github.com/enso-org/rust-lib/pull/32 being merged first.

### Checklist
Please include the following checklist in your PR:

~- [ ] The `CHANGELOG.md` was updated with the changes introduced in this PR.~
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
~- [ ] All code has automatic tests where possible.~
~- [ ] All code has been profiled where possible.~
~- [ ] All code has been manually tested in the IDE.~
- [x] All code has been manually tested in the "debug/interface" scene.
- [x] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [x] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

File: src/rust/ensogl/lib/core/src/data/function/traits.rs [hunk 1]
```diff
@@ -77,6 +77,7 @@ macro_rules! define_fn {
         }
 
         #[allow(non_snake_case)]
+        #[allow(clippy::manual_map)]
```
Comment (farmaazon) [inline src/rust/ensogl/lib/core/src/data/function/traits.rs:80]: Why do we allow? Is there any problem in fixing that? If yes, write a short comment why (so no one will ponder on that in the future).

File: src/rust/ensogl/lib/core/src/display/style/theme.rs [hunk 2]
```diff
@@ -303,10 +303,9 @@ impl Default for Manager {
 
 impl From<&style::Sheet> for Manager {
     fn from(style_sheet:&style::Sheet) -> Self {
-        let mut this = Self::default();
-        this.data    = Rc::new(RefCell::new(style_sheet.into()));
-        this.handles = default();
-        this
+        let data    = Rc::new(RefCell::new(style_sheet.into()));
+        let handles = default();
+        Self {data,handles,..default()}
```
Comment (farmaazon) [inline src/rust/ensogl/lib/core/src/display/style/theme.rs]: Have you checked if `handles` couldn't be in `..default()` as well?

File: src/rust/ensogl/lib/core/src/lib.rs [hunk 3]
```diff
@@ -8,20 +8,16 @@
...
 #![feature(const_type_id)]
 #![feature(drain_filter)]
 #![feature(fn_traits)]
-#![feature(overlapping_marker_traits)]
-#![feature(slice_patterns)]
 #![feature(specialization)]
 #![feature(trait_alias)]
 #![feature(try_trait)]
-#![feature(type_alias_impl_trait)]
+#![feature(marker_trait_attr)] // TODO needed?
```
Comment (farmaazon) [inline src/rust/ensogl/lib/core/src/lib.rs]: Remove feature or TODO.

File: src/rust/ensogl/lib/core/src/system/gpu/data/gl_enum.rs [hunk 4]
```diff
@@ -31,13 +31,13 @@ pub mod traits {
...
         /// Converts the current value to `GlEnum`.
-        fn into_gl_enum(&self) -> GlEnum;
+        fn to_gl_enum(&self) -> GlEnum;
     }
 
-    impl<T> IntoGlEnum for T where for<'a> &'a T:Into<GlEnum> {
-        fn into_gl_enum(&self) -> GlEnum {
+    impl<T> ToGlEnum for T where for<'a> &'a T:Into<GlEnum> {
+        fn to_gl_enum(&self) -> GlEnum {
             self.into()
```
Comment (farmaazon) [inline src/rust/ensogl/lib/core/src/system/gpu/data/gl_enum.rs:41]: Why this change? Is it something clippy complained about?

I'm just curious.

File: src/rust/ide/src/model/project.rs [hunk 5]
```diff
@@ -47,16 +47,19 @@ pub trait API:Debug {
...
 
     /// Creates a new execution context with given definition as a root; and registers the context
     /// for receiving update.
+    #[allow(clippy::needless_lifetimes)]
     fn create_execution_context<'a>
     (&'a self, root_definition:language_server::MethodPointer)
     -> BoxFuture<'a,FallibleResult<model::ExecutionContext>>;
 
     /// Set a new project name.
+    #[allow(clippy::needless_lifetimes)]
```
Comment (farmaazon) [inline src/rust/ide/src/model/project.rs]: Explain in a comment why we allow these needless lifetimes.

File: src/rust/ide/view/graph-editor/src/lib.rs [hunk 6]
```diff
@@ -5,19 +5,16 @@
 //! be of poor quality. Expect drastic changes.
 
 #![feature(associated_type_defaults)]
-#![feature(clamp)]
 #![feature(drain_filter)]
 #![feature(entry_insert)]
 #![feature(fn_traits)]
-#![feature(overlapping_marker_traits)]
+#![feature(marker_trait_attr)] // TODO[needed?]
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/lib.rs]: Remove feature or TODO

File: src/rust/lib/config/build.rs [hunk 7]
```diff
@@ -29,11 +28,18 @@ fn main() {
...
+pub struct Config {{
+{def}
+}}
+
+pub const CONFIG : Config = Config {{
+{inst}
+}};
+
+{vars}"#,
+    config_path=CONFIG_PATH,def=def,inst=inst,vars=vars);
```
Comment (farmaazon) [inline src/rust/lib/config/build.rs]: If we have foo=bar pairs, I would add a space after commas.

File: src/rust/lib/frp/src/io/keyboard.rs [hunk 8]
```diff
@@ -124,15 +124,14 @@ impl Key {
...
-            Self::Shift    (_) => false,
-            _                  => true
-        }
+        !matches!(self
+            , Self::Alt      (_)
+            | Self::AltGr    (_)
+            | Self::AltGraph (_)
+            | Self::Control  (_)
+            | Self::Meta     (_)
+            | Self::Shift    (_)
```
Comment (farmaazon) [inline src/rust/lib/frp/src/io/keyboard.rs]: I think we don't need it in separate lines.

File: src/rust/lib/system/web/src/lib.rs [hunk 9]
```diff
@@ -497,73 +495,16 @@ extern "C" {
...
-}
+// pub fn set_stdout() {
+//     let printer = Printer::new(_print, true);
+//     std::io::set_print(Some(Box::new(printer)));
+// }
+//
+// pub fn set_stdout_unbuffered() {
+//     let printer = Printer::new(_print, false);
+//     std::io::set_print(Some(Box::new(printer)));
+// }
```
Comment (farmaazon) [inline src/rust/lib/system/web/src/lib.rs]: The commented code should be removed.

**Review by farmaazon — APPROVED** [review]

File: src/rust/ensogl/lib/core/src/system/gpu/data/gl_enum.rs (hunk 4 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/system/gpu/data/gl_enum.rs:41]: The "into_*" and "to_*" methods are supposed to follow the convention about self type, see: https://rust-lang.github.io/rust-clippy/master/#wrong_self_convention

File: src/rust/ensogl/lib/core/src/data/function/traits.rs (hunk 1 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/data/function/traits.rs:80]: Commented.

File: src/rust/ensogl/lib/core/src/display/style/theme.rs (hunk 2 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/display/style/theme.rs]: Good catch!

File: src/rust/ensogl/example/src/lib.rs [hunk 10]
```diff
@@ -4,17 +4,14 @@
...
 #![feature(fn_traits)]
-#![feature(overlapping_marker_traits)]
 #![feature(specialization)]
 #![feature(trait_alias)]
-#![feature(type_alias_impl_trait)]
+#![feature(min_type_alias_impl_trait)]
 #![feature(unboxed_closures)]
-#![feature(vec_remove_item)]
-#![feature(weak_into_raw)]
+#![allow(incomplete_features)] // To be removed, see: https://github.com/enso-org/ide/issues/1559
```
Comment (wdanilo) [inline src/rust/ensogl/example/src/lib.rs:14]: I dont agree it sohuld be removed. This claim is too hard. I agree we might benefit from investigating further if min specializations will suit our case, but I don't agree that we NEED to remove it. See my comments in the linked issue to learn more.

File: src/rust/ensogl/lib/core/src/animation/easing.rs [hunk 11]
```diff
@@ -384,6 +384,8 @@ impl<T,F,OnStep,OnEnd> WeakAnimationLoop<T,F,OnStep,OnEnd> {
 
 /// Alias for `FixedFrameRateLoop` with specified step callback.
 pub type AnimationStep<T,F,OnStep,OnEnd> = animation::Loop<Step<T,F,OnStep,OnEnd>>;
+
+#[allow(missing_docs)]
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/animation/easing.rs]: Please add docs instead.

File: src/rust/ensogl/lib/core/src/animation/loops.rs [hunk 12]
```diff
@@ -144,6 +144,7 @@ where Callback : LoopCallback {
     }
 }
 
+#[allow(missing_docs)]
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/animation/loops.rs]: Please, add docs instead.

File: src/rust/ensogl/lib/core/src/animation/physics/inertia.rs [hunk 13]
```diff
@@ -681,6 +681,8 @@ impl<T,OnStep,OnStart,OnEnd> WeakAnimationLoop<T,OnStep,OnStart,OnEnd> {
 /// Alias for `FixedFrameRateLoop` with specified step callback.
 pub type FixedFrameRateAnimationStep<T,OnStep,OnStart,OnEnd> =
     animation::FixedFrameRateLoop<Step<T,OnStep,OnStart,OnEnd>>;
+
+#[allow(missing_docs)]
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/animation/physics/inertia.rs]: Add docs instead, please.

File: src/rust/ensogl/lib/core/src/data/color/space/conversion.rs [hunk 14]
```diff
@@ -352,9 +354,9 @@ mod tests {
                     let rgb  = Rgb::new(nr,ng,nb);
                     let lch  = Lch::from(rgb);
                     let rgb2 = Rgb::from(lch);
-                    let r2   = (rgb2.red   * 255.0) as i32;
-                    let g2   = (rgb2.green * 255.0) as i32;
-                    let b2   = (rgb2.blue  * 255.0) as i32;
+                    let _r2   = (rgb2.red   * 255.0) as i32;
+                    let _g2   = (rgb2.green * 255.0) as i32;
+                    let _b2   = (rgb2.blue  * 255.0) as i32;
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/data/color/space/conversion.rs]: alignment

File: src/rust/ensogl/lib/core/src/display/style/sheet.rs [hunk 15]
```diff
@@ -155,8 +155,12 @@ impl Debug for Expression {
 
 impl PartialEq for Expression {
     fn eq(&self, other:&Self) -> bool {
-        let same_args     = self.args == other.args;
-        let same_function = Rc::ptr_eq(&self.function,&other.function);
+        let same_args = self.args == other.args;
+        // The *const u8 cast discards the vtable to work around
+        // https://github.com/rust-lang/rust/issues/46139
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/display/style/sheet.rs]: Please extend the comment, so reading the comment would be enough to understand whats happening here. Lets use links only as "learn more" - the comment should be enough to get first grasp off whats going on.

File: src/rust/ensogl/lib/core/src/lib.rs (hunk 3 above)
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/lib.rs]: The comment was closed, but the code is still here.

File: src/rust/ensogl/lib/core/src/lib.rs [hunk 16]
```diff
@@ -8,20 +8,16 @@
...
+#![feature(min_type_alias_impl_trait)]
 #![feature(unboxed_closures)]
-#![feature(vec_remove_item)]
-#![feature(weak_into_raw)]
-
+#![allow(incomplete_features)] // To be removed, see: https://github.com/enso-org/ide/issues/1559
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/lib.rs:20]: as earlier - please apply my earlier comment to all such places.

File: src/rust/ensogl/lib/core/src/system/gpu/data/buffer.rs [hunk 17]
```diff
@@ -450,3 +451,31 @@ pub trait IsBuffer {
     fn vertex_attrib_pointer (&self, index:u32, instanced:bool);
     fn set_to_default        (&self, index:usize);
 }
+
+// Calls are not recursive, as inherent methods are preferred over methods provided by trait.
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/system/gpu/data/buffer.rs:455]: I dont understand this code. Why something like the default impl of e.g. `update` would ever be useful? Using the default one will cause recursion, right?

File: src/rust/ensogl/lib/core/src/system/gpu/data/buffer/item.rs [hunk 18]
```diff
@@ -180,7 +180,7 @@ impl Storable for f32 {
     fn slice_to_items_mut   (buffer: &mut [Self]) -> &mut [Self::Cell] { buffer }
 }
 
-impl<T:Storable<Cell=T>,R,C> Storable for MatrixMN<T,R,C>
+impl<T:Storable<Cell=T>,R,C> Storable for OMatrix<T,R,C>
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/system/gpu/data/buffer/item.rs:183]: Did they change the name to `OMatrix`? ...

File: src/rust/ensogl/lib/core/src/system/gpu/data/gl_enum.rs (hunk 4 above)
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/system/gpu/data/gl_enum.rs:41]: This is a good change. Im wondering why it was ever named `into_`

File: src/rust/ensogl/lib/core/src/system/gpu/data/uniform.rs [hunk 19]
```diff
@@ -176,20 +176,6 @@ impl<Value:Clone> {
...
-        self.value.upload_uniform(context,location);
-    }
-}
-
-impl<Value:UniformUpload> Uniform<Value> {
-    /// Uploads the uniform data to the provided location of the currently bound shader program.
-    pub fn upload(&self, context:&Context, location:&WebGlUniformLocation) {
-        self.rc.borrow().upload(context,location)
-    }
-}
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/system/gpu/data/uniform.rs:191]: why was it moved to a trait? Now the trait needs to be imported. Why is it better this way?

File: src/rust/ensogl/lib/text/src/buffer/view/selection.rs [hunk 20]
```diff
@@ -203,6 +203,7 @@ impl<T:Boundary> Selection<T> {
 
     /// Indicate whether this region should merge with the next.
     /// Assumption: regions are sorted (self.min() <= other.min())
+    #[allow(clippy::suspicious_operation_groupings)]
```
Comment (wdanilo) [inline src/rust/ensogl/lib/text/src/buffer/view/selection.rs:206]: Why this is needed? Can we fix it somehow without disabling clippy?

File: src/rust/ide/src/model/project.rs [hunk 21]
```diff
@@ -91,6 +94,12 @@ pub trait API:Debug {
     fn subscribe(&self) -> Subscriber<Notification>;
 }
 
+// Note: Needless lifetimes
+// ~~~~~~~~~~~~~~~~~~~~~~~~
+// Clippy complains about using explicit lifetimes, however the suggested change breaks compilation.
+// Apparently `BoxFuture` must be given an explicit lifetime, and this lifetime cannot be anonymous.
+// Thus, we silent the warning manually.
+
```
Comment (wdanilo) [inline src/rust/ide/src/model/project.rs]: Could you please report it to the clippy issue tracker and link it here?

File: src/rust/ide/src/tests.rs [hunk 22]
```diff
@@ -42,6 +42,13 @@ async fn failure_to_open_project_is_reported() {
         }]
     }));
     fixture.when_stalled_send_error(1,"Service error");
+    // FIXME [mwu]
+    //  For some reasons, the line below is needed, even though dropping the fixture does the very
+    //  same call. Apparently `run_until_stalled` must be called twice. This looks like an error,
+    //  and happens only for wasm targets.
+    //  This did not happen on the 2019 nightly.
+    //  Further investigation needed.
```
Comment (wdanilo) [inline src/rust/ide/src/tests.rs]: "the line below is needed" - the comment lacks the most important part - what happens without this line? Please add it there so the comment explains the whole world

File: src/rust/lib/code-builder/src/lib.rs [hunk 23]
```diff
@@ -3,6 +3,7 @@
 //! easily extendable in case other targets will be needed in the future.
 
 #![feature(specialization)]
+#![allow(incomplete_features)] // To be removed, see: https://github.com/enso-org/ide/issues/1559
```
Comment (wdanilo) [inline src/rust/lib/code-builder/src/lib.rs:6]: as earlier

File: src/rust/lib/frp/src/lib.rs [hunk 24]
```diff
@@ -144,6 +144,7 @@
 //!
 //! ```
 
+#![allow(incomplete_features)] // To be removed, see: https://github.com/enso-org/ide/issues/1559
```
Comment (wdanilo) [inline src/rust/lib/frp/src/lib.rs:147]: as earlier

File: src/rust/lib/system/web/src/lib.rs [hunk 25]
```diff
@@ -497,74 +495,6 @@ extern "C" {
...
-impl std::io::Write for Printer {
-    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
-        self.buffer.push_str(&String::from_utf8_lossy(buf));
-
-        if !self.is_buffered {
-            (self.printfn)(&self.buffer)?;
-            self.buffer.clear();
-
-            return Ok(buf.len());
-        }
```
Comment (wdanilo) [inline src/rust/lib/system/web/src/lib.rs:527]: Please provide `println!` mcro that works in the web (that uses `console.log` user the hood) in this PR, as part of our prelude. Otherwise, this will make development a lot harder and will cause trouble among all devs.

**Review by wdanilo — CHANGES_REQUESTED** [review]

**Comment by wdanilo** [comment]
Amazing job with it! I'm waiting like crazy for it to land. Really amazing job, Michal!

**Comment by wdanilo** [comment]
Also, Michal, please do this before merge:
image

File: src/rust/ensogl/example/src/lib.rs (hunk 10 above)
Comment (mwu-tow) [reply src/rust/ensogl/example/src/lib.rs:14]: To be discussed.

File: src/rust/ensogl/lib/core/src/lib.rs (hunk 3 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/lib.rs]: Truly puzzling, as IntelliJ for some reason doesn't show anything when I use "find all" for `marker_trait_attr`,

File: src/rust/ensogl/lib/core/src/system/gpu/data/buffer.rs (hunk 17 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/system/gpu/data/buffer.rs:455]: It serves only to redirect calls. `enum_dispatch` expects all variant types to implement the trait and explicitly uses it.

File: src/rust/ensogl/lib/core/src/system/gpu/data/buffer/item.rs (hunk 18 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/system/gpu/data/buffer/item.rs:183]: Like "Owned Matrix".
Old names are deprecated, likely due to stabilizing usize generic parameters.

File: src/rust/ensogl/lib/core/src/system/gpu/data/uniform.rs (hunk 19 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/system/gpu/data/uniform.rs:191]: `enum_dispatch` requires the trait to be implemented.

File: src/rust/ensogl/lib/text/src/buffer/view/selection.rs (hunk 20 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/text/src/buffer/view/selection.rs:206]: `(self.is_cursor() || other.is_cursor()) && other.min() == self.max();` looks suspicious apparently. It'd like min to be max or vice versa, as it expects that both `self` and `other` are checked for the same field.
If you have any neat ideas on refactoring, I'll apply. Otherwise, I want changes in this PR as trivial as possible, to avoid regressions.

File: src/rust/ide/src/tests.rs (hunk 22 above)
Comment (mwu-tow) [reply src/rust/ide/src/tests.rs]: The test fails.

File: src/rust/ensogl/lib/core/src/system/gpu/data/buffer.rs (hunk 17 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/system/gpu/data/buffer.rs:455]: Write why needed in a comment

File: src/rust/ensogl/example/src/lib.rs (hunk 10 above)
Comment (mwu-tow) [reply src/rust/ensogl/example/src/lib.rs:14]: Agreed to left as-is.

File: src/rust/ensogl/lib/core/src/lib.rs (hunk 16 above)
Comment (mwu-tow) [reply src/rust/ensogl/lib/core/src/lib.rs:20]: Agreed to left as-is.

File: src/rust/ide/src/model/project.rs (hunk 21 above)
Comment (mwu-tow) [reply src/rust/ide/src/model/project.rs]: I've been unable to provide self-contained repro.
Clippy is correct, it seems that `automock`-generated code causes the issue.
The comment has been updated.

File: src/rust/ide/src/tests.rs (hunk 22 above)
Comment (mwu-tow) [reply src/rust/ide/src/tests.rs]: Comment has been updated.

File: src/rust/lib/code-builder/src/lib.rs (hunk 23 above)
Comment (mwu-tow) [reply src/rust/lib/code-builder/src/lib.rs:6]: Agreed to left as-is.

File: src/rust/lib/frp/src/lib.rs (hunk 24 above)
Comment (mwu-tow) [reply src/rust/lib/frp/src/lib.rs:147]: Agreed to left as-is.

File: src/rust/lib/system/web/src/lib.rs (hunk 25 above)
Comment (mwu-tow) [reply src/rust/lib/system/web/src/lib.rs:527]: Non trivial, as we have no proper means of hiding the default `println!`, adding a new one to prelude just makes them ambiguous.
Also, this cannot be part of this PR, since prelude belongs to a separate repo.
To be discussed how to handle this.

File: docs/CONTRIBUTING.md [hunk 26]
```diff
@@ -52,7 +52,7 @@ application icon if built on Linux or Windows due to non-trivial icon generation
 on these platforms. To develop the source code you will need the following
 setup:
 
-- **The Rust Toolchain (nightly-2019-11-04)**
+- **The Rust Toolchain (nightly-2021-05-12)**
```
Comment (MichaelMauderer) [inline docs/CONTRIBUTING.md:55]: This also needs updating in the commands in the section below. (GH won't let me comment in the line directly as it has not been changed).
`rustup toolchain install nightly-2019-11-04`
to 
`rustup toolchain install nightly-2021-05-12`

Comment (mwu-tow) [reply docs/CONTRIBUTING.md:55]: Is fixed

**Review by wdanilo — APPROVED** [review]
